modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EChain.java

```diff
@@ -233,6 +233,7 @@ private void analyzeWrite(final CompilerSettings settings, final Definition defi
         if (last instanceof ADefLink) {
             final ADefLink lastDef = (ADefLink) last;
             expression.analyze(settings, definition, variables);
+            // TODO: does it make more sense to just re-use last.after instead of using storeValueType?
```

Comment:
I had something similar before, maybe revert back to using ALink.after for that - I was not sure if the operand should always have the same type like the after value. In any case I would keep the superclass ADefLink to make them easier to "detect" in those if statemments, although the ADefLink class is empty afterwards.